File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = 100
    graph = [[] for _ in range(n+1)]
    
    for v1, v2 in wires:
        graph[v1].append(v2)
        graph[v2].append(v1)
    
    def bfs():
        count = 1
        visited = [False] * (n+1)
        queue = deque()
        queue.append(1)
        visited[1] = True
        while queue:
            cur = queue.popleft()
            for nxt in graph[cur]:
                if not visited[nxt]:
                    visited[nxt] = True
                    queue.append(nxt)
                    count += 1
        return abs(count - (n-count)) 
    
    for i in range (len(wires)):
        v1, v2 = wires[i]
        graph[v1].remove(v2)
        graph[v2].remove(v1)
        temp = bfs()
        answer = min(answer, temp)
        graph[v1].append(v2)
        graph[v2].append(v1)
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> graph;
int answer;

int dfs(int cur, int parent, int n) {
    int size = 1;
    
    for (auto nxt : graph[cur]) {
        if (nxt == parent) continue;
        int child_size = dfs(nxt, cur, n);
        int diff = abs(child_size - (n - child_size));
        answer = min(answer, diff);
        size += child_size;
    }
    
    return size;
}

int solution(int n, vector<vector<int>> wires) {
    answer = INT_MAX;
    graph.resize(n+1);
    
    for (auto v : wires) {
        int v1 = v[0];
        int v2 = v[1];
        graph[v1].push_back(v2);
        graph[v2].push_back(v1);
    }
    
    dfs(1, 0, n);
    
    return answer;
}
